components/chainhook-cli/src/service/tests/mod.rs

```diff
@@ -559,89 +271,442 @@ fn assert_streaming_status(status: PredicateStatus) {
-    let block = create_stacks_tsv_block(height, burn_block_height);
-    let serialized_block = serde_json::to_string(&block).unwrap();
-    let _block_again: NewBlock = serde_json::from_str(&serialized_block).unwrap();
-    let client = reqwest::Client::new();
-    let _res = client
-        .post(format!("http://localhost:{port}/new_block"))
-        .header("content-type", "application/json")
-        .body(serialized_block)
-        .send()
+fn assert_interrupted_status(status: PredicateStatus) {
+    match status {
+        PredicateStatus::Interrupted(_) => {}
+        _ => panic!("expected Interrupted status, found {:?}", status),
+    }
+}
+
+fn setup_chainhook_service_ports() -> Result<(u16, u16, u16, u16, u16), String> {
+    let redis_port = get_free_port()?;
+    let chainhook_service_port = get_free_port()?;
+    let stacks_rpc_port = get_free_port()?;
+    let stacks_ingestion_port = get_free_port()?;
+    let bitcoin_rpc_port = get_free_port()?;
+    Ok((
+        redis_port,
+        chainhook_service_port,
+        stacks_rpc_port,
+        stacks_ingestion_port,
+        bitcoin_rpc_port,
+    ))
+}
+
+async fn await_new_scanning_status_complete(
+    uuid: &str,
+    chainhook_service_port: u16,
+) -> Result<(), String> {
+    let mut attempts = 0;
+    loop {
+        match get_predicate_status(uuid, chainhook_service_port).await? {
+            PredicateStatus::New | PredicateStatus::Scanning(_) => {
+                attempts += 1;
+                if attempts == 10 {
+                    return Err(format!("predicate stuck in new/scanning status"));
+                }
+                sleep(Duration::new(1, 0));
+            }
+            _ => break Ok(()),
+        }
+    }
+}
+
+async fn setup_stacks_chainhook_test(
+    starting_chain_tip: u64,
+) -> (Child, String, u16, u16, u16, u16) {
+    let (
+        redis_port,
+        chainhook_service_port,
+        stacks_rpc_port,
+        stacks_ingestion_port,
+        bitcoin_rpc_port,
+    ) = setup_chainhook_service_ports().unwrap_or_else(|e| panic!("test failed with error: {e}"));
+
+    let mut redis_process = start_redis(redis_port)
         .await
-        .unwrap()
-        .text()
+        .unwrap_or_else(|e| panic!("test failed with error: {e}"));
+
+    let (working_dir, tsv_dir) = create_tmp_working_dir().unwrap_or_else(|e| {
+        flush_redis(redis_port);
+        redis_process.kill().unwrap();
+        panic!("test failed with error: {e}");
+    });
+
+    let logger = hiro_system_kit::log::setup_logger();
+    let _guard = hiro_system_kit::log::setup_global_logger(logger.clone());
+    let ctx = Context {
+        logger: Some(logger),
+        tracer: false,
+    };
+
+    write_stacks_blocks_to_tsv(starting_chain_tip, &tsv_dir).unwrap_or_else(|e| {
+        std::fs::remove_dir_all(&working_dir).unwrap();
+        flush_redis(redis_port);
+        redis_process.kill().unwrap();
+        panic!("test failed with error: {e}");
+    });
```

Comment:
@lgalabru  I repeatedly use this pattern a lot, but I'd really like to be able to use something like: 

```suggestion
    write_stacks_blocks_to_tsv(starting_chain_tip, &tsv_dir).map_err(|e| cleanup(e.to_string(), redis_process,redis_port,working_dir))?;
```

My problem is that passing `redis_process` into a function moves it, so it can't be called by `cleanup` more than once. This doesn't really matter - cleanup always panics and can only be called once. However, the borrow checker doesn't know this. 

The obvious workaround is to use `redis_process.clone()`, but its type is `std::process::Child`, which we obviously cannot alter and doesn't have `Clone` implemented.

Do you know of a way that I can simplify this pattern and reduce the number of lines of code? Maybe we could pair on this.